pkg/document/manager.go

```diff
@@ -36,31 +48,112 @@ func WithNewDocumentFunc(newDocFunc NewDocumentFunc) ManagerOpt {
 	}
 }
 
+func WithReadDocumentFunc(readDocFunc ReadDocumentFunc) ManagerOpt {
+	return func(manager *Manager) {
+		manager.readDocFunc = readDocFunc
+	}
+}
+
+func ReadDocument(u uri.URI) (contents []byte, err error) {
+	fn, err := filename(u)
+	if err != nil {
+		return nil, err
+	}
+	return os.ReadFile(fn)
+}
+
+func filename(u uri.URI) (fn string, err error) {
+	defer func() {
+		// recover from non-file URI in uri.Filename()
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+	return u.Filename(), err
+}
+
+func canonicalURI(u uri.URI, base uri.URI) uri.URI {
+	fn, err := filename(u)
+	if err != nil {
+		return u
+	}
+	if !filepath.IsAbs(fn) && base != "" {
+		basepath, err := filename(base)
+		if err != nil {
+			return u
+		}
+		fn = filepath.Join(basepath, fn)
+	}
+	fn, err = filepath.EvalSymlinks(fn)
+	if err != nil {
+		return u
+	}
+	return uri.File(fn)
+}
+
+func (m *Manager) Initialize(params *protocol.InitializeParams) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if len(params.WorkspaceFolders) > 0 {
+		m.root = uri.URI(params.WorkspaceFolders[0].URI)
+	} else {
+		dir, err := os.Getwd()
+		if err == nil {
+			m.root = uri.File(dir)
+		}
+	}
+}
+
 // Read returns the contents of the file for the given URI.
 //
 // If no file exists at the path or the URI is of an invalid type, an error is
 // returned.
-func (m *Manager) Read(uri uri.URI) (Document, error) {
+func (m *Manager) Read(ctx context.Context, u uri.URI) (doc Document, err error) {
 	m.mu.Lock()
-	defer m.mu.Unlock()
-	if doc, ok := m.docs[uri]; ok {
-		return doc.Copy(), nil
+	defer func() {
+		if err == nil {
+			// Always return a copy of the document
+			doc = doc.Copy()
+		}
+		m.mu.Unlock()
+	}()
+	u = canonicalURI(u, m.root)
+
+	// TODO(siegs): check staleness for files read from disk?
```

Comment:
Wondering if this should be handled now or in a future PR -- as written this caches documents and they won't be updated unless separately edited in the same editor that is talking to the current LSP server. If you make edits outside of the LSP client editor, they won't be reflected.